### ISSUE TITLE
Shiki language loading is slow

### ISSUE THREAD

**Issue by fweth** [expressive-code/expressive-code#256]
```
"astro": "4.16.6",
"astro-expressive-code": "0.37.1"
```

Create a new Astro project and add something like
```html
<Code code={"test"} lang="txt" />
```
to the index page. Then run `npm run build`, it takes between 6 and 7 seconds for the index page only to build (not the whole build process). When code component is included on multiple pages, only the first one takes long. Also noticable in the dev server when navigating to a page containing the code component. Tested locally (Linux) and on StackBlitz.

**Comment by hippotastic** [comment]
I can't reproduce this on any of the machines I tested. Happy to reopen if you can provide more info & a reproduction.

**Comment by hippotastic** [comment]
Oh, I forgot to mention one idea: If the system you're running this on is particularly slow or limited, I believe it's just the upstream bundling that might be taking longer.

You might be interested in trying the plugin I posted in another issue to cut down the primary source of content to bundle - unneeded Shiki languages & themes. See: https://github.com/expressive-code/expressive-code/issues/251#issuecomment-2415195839

**Comment by fweth** [comment]
Here is some StackBlitz: https://stackblitz.com/edit/withastro-astro-3vcdd4

With the line including `<Code>`, the output is

```
 generating static routes 
17:38:30 ▶ src/pages/index.astro
17:38:30   └─ /index.html (+8.25s)
17:38:38 ✓ Completed in 8.41s.
```

and without that line I get

```
 generating static routes 
17:38:52 ▶ src/pages/index.astro
17:38:52   └─ /index.html (+39ms)
17:38:52 ✓ Completed in 119ms.
```

Those numbers are from StackBlitz. On my local system, it's similar:

```
 generating static routes 
17:35:43 ▶ src/pages/index.astro
17:35:43   └─ /index.html (+6.36s)
17:35:50 ✓ Completed in 6.37s.
```

vs.

```
 generating static routes 
17:36:21 ▶ src/pages/index.astro
17:36:21   └─ /index.html (+6ms)
17:36:21 ✓ Completed in 12ms.
```

**Comment by hippotastic** [comment]
Can you please try my plugin I mentioned? This is probably an upstream bundling issue I can do nothing about.

**Comment by fweth** [comment]
Like this?

```js
import { defineConfig } from "astro/config";

import expressiveCode from "astro-expressive-code";

// https://astro.build/config
export default defineConfig({
  integrations: [
    expressiveCode(),
    {
      name: "trim-shiki",
      hooks: {
       ...
      },
    },
  ],
});
```

It didn't change anything. Also, if it is about `shiki`, shouldn't the same problem persist for `md` files as well, with code snippets?

**Comment by hippotastic** [comment]
I just had a look and it is indeed due to Shiki. It's the mere loading of Shiki languages that is so slow.

You can see the effect of disabling Shiki with the following config:
```ts
export default defineConfig({
  integrations: [
    expressiveCode({
      shiki: false,
    }),
  ],
});
```

After just this change, the generation times are a lot faster on StackBlitz:

```
 generating static routes 
23:17:44 ▶ src/pages/index.astro
23:17:44   └─ /index.html (+169ms)
23:17:44 ✓ Completed in 319ms.
```

Please note that when comparing timings between code blocks in `.md` files and the `<Code>` component, the entire build time will always be the same, it's just a different phase that will be slowed down. When using a fenced code block in an `.md` file, the language loading will happen in the "Building static entrypoints..." phase. When using the `<Code>` component, the loading will happen in the "generating static routes" phase.

Sadly, my bundle trimming plugin really doesn't work in this situation, it only seems to help in SSR scenarios. I think I'll need to revisit the language loading and try to enable the lazy loading of the actually used languages again. Last time I used this, it triggered a race condition inside Shiki which caused builds to hang. I hope this has been fixed in the meanwhile.

**Comment by fweth** [comment]
Thanks a lt for looking into it! Should we close this issue and I'll try to open a new Issue for Shiki?

**Comment by hippotastic** [comment]
No, I think we can keep this one open and treat it as a feature request to allow more granular language loading. I'll look into it!

**Comment by fweth** [comment]
Thank you! I should add that the delay is also very noticable in the dev server, when you navigate to a route which has the `<Code>` component from a route which doesn't have it. This is what was actually bugging me, not the build times. So maybe the language loading could be done when starting the dev server and the results being cached or something, so you don't have the delay during navigation?

**Comment by hippotastic** [comment]
This should be fixed in version 0.38.0, which I just released to NPM. The languages are being loaded on demand now, significantly reducing the overhead during dev server startup & page loading, as well as speeding up the build and memory usage.

Thanks again for reporting and helping me narrow down the cause!
